DOC: Fix Multithreaded Generation example docs

fzeiser opened the pull request:
Working on #15365
The code runs now, but I get quite different runtimes than in the example. Thus it may need some more work.

Timing tests are also from the original documentation (but rerun here)

1. using multiple threads
``` python
print(mrng.threads)
%timeit mrng.fill()
# 4
# 52.6 ms ± 508 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

2. single threaded directly
``` python
values = np.empty(10000000)
rg = default_rng(PCG64())
%timeit rg.standard_normal(out=values)
# 143 ms ± 1.14 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

3. single threaded directly, but do not use an existing array.
``` python
rg = default_rng(PCG64())
%timeit rg.standard_normal(10000000)
# 148 ms ± 590 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

So following quote would not make sense on my machine.
> 
> The gains are substantial and the scaling is reasonable even for large that
> are only moderately large.  The gains are even larger when compared to a call
> that does not use an existing array due to array creation overhead.

Can someone confirm this? Or did I implement something wrong?

Commit:
DOC, BUG: Fixes Multithreaded Generation docs

Working on #15365
The code runs now, but I get quite different runtimes than
in the example. May need some more work

rossbar reviewed (changes requested):
This is a really nice catch @fzeiser, I think you're assessment is spot on: the issue stems from the fact that the `rg` in the `MultithreadedRNG` are `BitGenerator`s rather than `Generator`s.

I'd suggest a couple changes that would make things a bit more concise:
 1. Don't replace `Generator` with `default_rng` - the change you've made will work, but `default_rng` is intended to be a convenience function. `Generator(PCG64())` is equivalent to `default_rng()`.
 2. Do away with the extra list-comp at the end. The same functionality can be achieved by modifying ~with~ the for loop:
```python
for _ in range(0, threads-1):
        new_bg = last_bg.jumped()
        self._random_generators.append(Generator(new_bg))
        last_bg = new_bg
```

I agree with you that replacing the instances of `rg` with `bg` is appropriate to make the distinction between `BitGenerator` and `Generator` more clear.

rossbar commented:
Re: Performance, I see the following:

```python
>>> from numpy.random import Generator, PCG64
>>> n = int(1e7)

>>> # Single thread
>>> %timeit rng.standard_normal(n)
95.6 ms ± 81.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> # Single thread, pre-allocated out ary
>>> values = np.empty(n)
>>> %timeit rg.standard_normal(out=values)
87.8 ms ± 94.9 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> # Multiple threads
>>> mrng = MultithreadedRNG(n, seed=0, threads=4)
>>> print(mrng.threads)
4
>>> %timeit mrng.fill()
24 ms ± 1.4 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The relative timing values seem consistent with those in the original doc.

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -19,24 +19,26 @@ seed will produce the same outputs.
 
 .. code-block:: ipython
 
-    from numpy.random import Generator, PCG64
```
Leave as `Generator` - reduces the number of changes.

fzeiser replied:
I've changed it as requested. It's a bit confusing for me though that the documentation of `Generator` has the  `See also` section saying:

 ```
See also

default_rng
    Recommended constructor for Generator.
```

mattip replied:
would changing Recommended -> Convenience help?

rossbar replied:
Maybe something like:

```
See also

default_rng
    Convenience constructor for Generator with PCG64 BitGenerator.
```

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -19,24 +19,26 @@ seed will produce the same outputs.
+                new_bg = last_bg.jumped()
+                self._random_generators.append(new_bg)
+                last_bg = new_bg
+            self._random_generators = [default_rng(bg) for bg in
```
This list-comp can be gotten rid of by changing two lines above:
```python
self._random_generators.append(Generator(new_bg))
```

Commit:
DOC: changes after PR review

Commit:
DOC: Changing input/output lines to sphinx standard

For ipython example: Adding `Out[i]` where i is the input cell number
and correcting vissual multiline inputs

rossbar reviewed (approved):
Looks good to me; I ran this locally and the results were consistent with the original doc.

seberg commented:
Thanks Ross and @fzeiser, I think we can put this in. @bashtage do you like this? The only thing I am worried about is where and how this is advertised. Having an example of the `jumped` API is good, but I would prefer if the `SeedSequence` API is more prominent in the documentation.

bashtage commented:
LGMT. This page looks like it wasn't updated to reflect the changes as randomgen was moved in.

bashtage commented:
> Can someone confirm this? Or did I implement something wrong?

IIRC large array allocation speed can meaningfully vary a lot depending on OS. But I think the sentence can go.

bashtage commented:
One thing I would say is that `jumped` is not the preferred method.  The best method looks like

```
seq = SeedSequence(2318918410984098328420)  # good entropy
bgs = [Generator(PCG64(sub_seq)) for sub_seq in seq.spawn(10)]
```

Or 
```
entropy = 2318918410984098328420
bgs = [Generator(PCG64(sub_seq)) for sub_seq in SeedSequence(entropy).spawn(10)]
```

fzeiser commented:
How about we change the rg generation part below? The 

``` py
            self._random_generators = [Generator(bg)]
            last_bg = bg
            for _ in range(0, threads-1):
                new_bg = last_bg.jumped()
                self._random_generators.append(Generator(new_bg))
                last_bg = new_bg
```

including the "advertisement":
_The underlying BitGenerator is PCG64 which is fast, has a long period and supports using PCG64.jumped to return a new generator while advancing the state._

to 

> ```py
> entropy = 2318918410984098328420
> bgs = [Generator(PCG64(sub_seq)) for sub_seq in SeedSequence(entropy).spawn(10)]
> ```

The reason I came across this side and I think that it is interesting, is that it gives a nice example on how to use multithreading to work on a single array.  To me, it wasn't important on whether this would be due to jump, or anything else.

bashtage commented:
Absolutely.  It should be written to show what is considered the best practice.  SeedSequence can be used with an BasicGen while jumped is not supported by all.  You should remove referenced to jumping, or make a small note towards the end that jumped is an alternative which uses disjoint blocks of the same sequence that are sufficiently far apart that they will never overlap.

peteroupc commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -61,40 +61,41 @@ seed will produce the same outputs.
+    Out[5]: 99.6 ms ± 222 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
 
 The gains are substantial and the scaling is reasonable even for large that
 are only moderately large.  The gains are even larger when compared to a call
```
Shouldn't this say "reasonable even for arrays that..."?

rossbar replied:
Good catch, feel free to make the edit

rossbar commented:
@fzeiser Looks like the right approach; it's nice that it makes the code simpler as well! I'll take a look at whatever you push

Thanks @seberg and @bashtage for the input - I wasn't aware that one method was preferred over the other

bashtage commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -26,17 +26,17 @@ seed will produce the same outputs.
-                new_rg = last_rg.jumped()
-                self._random_generators.append(new_rg)
-                last_rg = new_rg
+                new_bg = last_bg.jumped()
```
I though we were going to get rid of jumped. 
I think this could be 

```
    class MultithreadedRNG:
        def __init__(self, n, seed=None, threads=None):
            seq = SeedSequence(seed)
            if threads is None:
                threads = multiprocessing.cpu_count()
            self.threads = threads

            self._random_generators = [Generator(PCG64(s)) for s in seq.spawn(threads) ]
...
```

fzeiser replied:
done

bashtage reviewed (changes requested):
I think removing `jumped` in favor of using `SeedSequence` should be implemented.

Commit:
MAINT: Replace `jumped` by `SeedSequence` in multithreading snippet

After PR review, decided to replace `jumped` by `SeedSequence` as it is a more general method. Adopted the introduction to reflect this.

Also added a section on the reproducibility for different machines and number of threads + an example.

fzeiser reviewed (commented):
Thanks a lot for all the input :smile:. Submitted the fixes in bb579f3. 

I figured that I should also add a sentence  and example on that fact that using the default argument for `threads` will lead to different results on different machines. This is often not desirable. Do you have a good idea on how we could mitigate this effect?
Obviously, if one used `jumped`, one could calculate how long one needs to jump if an array of size `n` is filled by `m` processors. But now we said that we don't want to have `jumped`.

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -12,31 +12,27 @@ these requirements.
-                self._random_generators.append(Generator(new_bg))
-                last_bg = new_bg
+            seq = SeedSequence(seed)
+            self._random_generators = [Generator(PCG64(s)) for s in seq.spawn(threads)]
```
Just a style thing: consider breaking up the list comp so it stays under 80-char per line:

```python
self._random_generators = [
    Generator(PCG64(s)) for s in seq.spawn(threads)
]
```

rkern replied:
Since we've changed from using `.jumped()` to the `BitGenerator`-agnostic solution, I would recommend using `default_rng(s)` instead of `Generator(PCG64(s))`.

fzeiser replied:
Thanks for the input. I'll change it accordingly.

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -107,3 +103,11 @@ that does not use an existing array due to array creation overhead.
 
     Out[6]: 125 ms ± 309 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+
+Note that if `threads` is not set by the user, it will be determined by the architecture
```
I recommend being a bit more explicit here by replacing "the architecture" with `multiprocessing.cpu_count`.

fzeiser replied:
Sounds good! I wasn't sure on how explicit I should be.

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -107,3 +103,11 @@ that does not use an existing array due to array creation overhead.
     Out[6]: 125 ms ± 309 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
+
+Note that if `threads` is not set by the user, it will be determined by the architecture
+
```
I'm not sure this addition is necessary - the code defining `MultithreadedRNG` is pretty clear and a user should be able to grok how the number of threads is determined. Since the example doesn't highlight any of the `np.random` machinery itself I'd lean towards excluding this code-block.

fzeiser replied:
I would like to keep a reference in there. When I started working with multithreading and random generators, I stepped in the "common" trap of not thinking about what would happen to the random states. I think it is a common enough problem that (new) users are not aware of this, see eg. the discussion in #9650. What would you the others think @bashtage?

rossbar replied:
Oops, I must have put the wrong comment in the wrong bubble - I agree this can stay, though what I meant to suggest was to replace "the architecture" with "multiprocessing.cpu_count".

rossbar commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -12,31 +12,27 @@ these requirements.
 This example makes use of Python 3 :mod:`concurrent.futures` to fill an array
 using multiple threads.  Threads are long-lived so that repeated calls do not
 require any additional overheads from thread creation. The underlying
-BitGenerator is `PCG64` which is fast, has a long period and supports
```
@bashtage thoughts on the wording here?

fzeiser replied:
Oups, yes, this isn't ideal! As I will use `default_rng` I might just delete this sentence altogether.

rossbar reviewed (commented):
Just a couple suggestions; no show-stoppers so take it or leave it - thanks for taking the lead on these changes!

rossbar commented:
Since we now have relevant conversations happening in multiple PRs (#15391), here's a quick summary of what should be done to make this mergeable (as I understand it):

 * Address the three comments from the previous review. These are just suggestions so take-them-or-leave-them
 * Switch back to `default_rng()` instead of the `Generator(BitGenerator))` syntax pursuant @rkern 's explanation in #15391 
 * Rebase on master - this should fix the circleci failure, which is unrelated to the changes here.

@fzeiser hopefully I haven't worn you down too much with all the back and forth!

fzeiser commented:
Sorry for opening the new issue (#15391), obviously I'm not yet used to the workflow here and what should be forked out in another conversation. Thanks for making it so easy to follow along though, @rossbar.

I'll wait with changes to the following part until I will hear back from @rossbar /others:
[what's the easiest way to link to a conversation in a code review?]

> Note that if `threads` is not set by the user, it will be determined by the architecture
> 
> .. code-block:: ipython
>     In [7]: # simulate the behavior for `threads=None`, if the machine had only one thread
>        ...: mrng = MultithreadedRNG(10000000, seed=12345, threads=1)
>        ...: print(mrng.values[-1])
>     Out[7]: 1.1800150052158556

Commit:
MAINT: Changes after PR review

- Switch back to default_rng() instead of the Generator(BitGenerator)) syntax pursuant @rkern 's explanation in #15391
- small fixes

rossbar reviewed (approved):
Looks good to me, my only remaining comment is that I think `PCG64` can be removed entirely since `default_rng` uses it by default. Calls like `default_rng(PCG64())` on line 91 and 102 could be replaced with `default_rng()` and `PCG64` would no longer need to be imported.

Thanks for all the hard work @fzeiser

fzeiser commented:
@rossbar: Sorry, that was my intention. But then search-and-replace hit me in the back.

Commit:
MAINT: Delete unneccesary reference to PCG64

seberg commented:
Thanks, the test failure (in circleCI) is real:
```
/home/circleci/repo/doc/source/reference/random/multithreading.rst:110: WARNING: Error in "code-block" directive:
maximum 1 argument(s) allowed, 26 supplied.

.. code-block:: ipython
    In [7]: # simulate the behavior for `threads=None`, if the machine had only one thread
       ...: mrng = MultithreadedRNG(10000000, seed=12345, threads=1)
       ...: print(mrng.values[-1])
    Out[7]: 1.1800150052158556
Makefile:178: recipe for target 'html-build' failed
```
My best (quick guess) is that it has incorrect indentation?

Commit:
MAINT: Added newline finxing codeblock error in docs

Trying to see whether this fixes circleci error

seberg commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -11,32 +11,28 @@ these requirements.
 .. code-block:: ipython
 
-    from numpy.random import Generator, PCG64
+    from numpy.random import default_rng, PCG64, SeedSequence
```
```suggestion
    from numpy.random import default_rng, SeedSequence
```

fzeiser replied:
Thanks for catching this ;P!

seberg commented on doc/source/reference/random/multithreading.rst:
```diff
@@ -61,48 +57,59 @@ seed will produce the same outputs.
-    In [2]: mrng = MultithreadedRNG(10000000, seed=0)
-    ...: print(mrng.values[-1])
-    0.0
+    In [2]: mrng = MultithreadedRNG(10000000, seed=12345)
```
@rkern I was wondering about `seed=...` best practices for documentation (not necessarily here), do you think this is fine, or should we write such documentation in the form of `seed=0x...`. I have probably asked this before, but...

rkern replied:
Good question. Avoiding `seed=0` is important, IMO. Thanks, @fzeiser! Grabbing a random uint64 to use for a seed in the examples might help encourage safer practices. That is, the example writer executes something like the following in their Python shell

```
>>> hex(np.random.default_rng().integers(1<<64, dtype=np.uint64))
'0xdf05c6f9894cea6b'
```

Then use that value in their example:

```
"""
Examples
--------
In [2]: mrng = MultithreadedRNG(10000000, seed=0xdf05c6f9894cea6b)
"""
```

But I'm not going to complain too much about using `12345`. That does help communicate "this is a dummy value", so I get why example writers use it. I'm not sure that seeing `seed=12345` in an example influences too many people to use `seed=12345` in their own code.

seberg replied:
Well, I am more thinking about pushing best practices without actually forcing too much on users. I.e. we could make examples generally have a comment:
`# seed found using: hex(np.random.default_rng().integers(1<<64, dtype=np.uint64))`
(or anything prettier/shorter if we can think of it)

seberg replied:
If you think it is good practice, I would be happy with adding:
```
def create_unique_seed():
    return np.random.default_rng().integers(1<<64, dtype=np.uint64)
    # or maybe: return np.random.SeedSequence().entropy 
```
just so that such a comment can be short.

rkern replied:
I think we talked about having such a function back in the old mega-thread but held off on it to get some practice with the new system and `SeedSequence` in particular. Let's think about it for 1.19.

seberg reviewed (approved):
LGTM, nitpicking by now. Since Ross seems happy, I am happy to put this in as is even, ping me tomorrow or so if I forget.

Commit:
MAINT: Clean up in imports of code snippet

seberg commented:
Thanks @fzeiser, Ross and everyone else.
